File: go/pkg/lifecycle/manager.go
package lifecycle

// -----------------------------------------------------------------------------
// ESSENTIAL PROCESS:
// Manages application lifecycle, OS signal traps (SIGINT, SIGTERM), and orderly
// LIFO execution of registered graceful shutdown hooks.
//
// DATA FLOW:
// OS Signals / Context Cancel -> Manager.Wait() -> LIFO Cleanup Execution -> Process Exit
//
// KEY PARAMETERS:
// - cleanups: Slice of registered cleanupHook instances (executed LIFO).
// - Logger: Universal Logger instance for audit logging.
// -----------------------------------------------------------------------------

import (
	"context"
	"os"
	"os/signal"
	"syscall"

	"github.com/Bastien-Antigravity/microservice-toolbox/go/pkg/logger"
)

// ShutdownFunc is a function called during graceful shutdown.
type ShutdownFunc func() error

type cleanupHook struct {
	name string
	fn   ShutdownFunc
}

// Manager handles application lifecycle and graceful shutdown.
type Manager struct {
	cleanups []cleanupHook
	Logger   logger.Logger
}

// NewManager creates a new lifecycle manager with default logging.
func NewManager() *Manager {
	return NewManagerWithLogger(nil)
}

// NewManagerWithLogger creates a new lifecycle manager with an explicit logger.
func NewManagerWithLogger(l logger.Logger) *Manager {
	return &Manager{
		cleanups: make([]cleanupHook, 0),
		Logger:   logger.EnsureSafeLogger(l),
	}
}

// Register adds a named cleanup function to the shutdown sequence.
func (m *Manager) Register(name string, fn ShutdownFunc) {
	m.cleanups = append(m.cleanups, cleanupHook{name: name, fn: fn})
}

// Wait blocks until a SIGINT or SIGTERM is received or context is cancelled,
// then executes all registered cleanups in reverse registration order (LIFO).
func (m *Manager) Wait(ctx context.Context) {
	stop := make(chan os.Signal, 1)
	signal.Notify(stop, syscall.SIGINT, syscall.SIGTERM)

	select {
	case sig := <-stop:
		m.Logger.Info("Lifecycle: Received signal %v. Initiating graceful shutdown...", sig)
	case <-ctx.Done():
		m.Logger.Info("Lifecycle: Context canceled. Initiating graceful shutdown...")
	}

	// Execute cleanups in reverse order (LIFO)
	for i := len(m.cleanups) - 1; i >= 0; i-- {
		hook := m.cleanups[i]
		m.Logger.Info("Lifecycle: Executing cleanup hook [%s]...", hook.name)
		if err := hook.fn(); err != nil {
			m.Logger.Error("Lifecycle: Cleanup hook [%s] failed: %v", hook.name, err)
		}
	}
	m.Logger.Info("Lifecycle: Clean shutdown completed.")
}

File: rust/src/lifecycle/manager.rs
use std::sync::Arc;
use tokio::sync::Mutex;
use std::future::Future;
use std::pin::Pin;
use crate::utils::logger::{Logger, ensure_safe_logger};

pub type CleanupResult = Result<(), Box<dyn std::error::Error + Send + Sync>>;
pub type ShutdownFunc = Box<dyn Fn() -> Pin<Box<dyn Future<Output = CleanupResult> + Send>> + Send + Sync>;

struct CleanupEntry {
    name: String,
    func: ShutdownFunc,
}

pub struct LifecycleManager {
    cleanups: Arc<Mutex<Vec<CleanupEntry>>>,
    logger: Arc<dyn Logger>,
}

impl LifecycleManager {
    pub fn new(logger: Option<Arc<dyn Logger>>) -> Self {
        Self {
            cleanups: Arc::new(Mutex::new(Vec::new())),
            logger: ensure_safe_logger(logger),
        }
    }

    pub async fn register<F, Fut>(&self, name: &str, f: F)
    where
        F: Fn() -> Fut + Send + Sync + 'static,
        Fut: Future<Output = CleanupResult> + Send + 'static,
    {
        let mut cleanups = self.cleanups.lock().await;
        cleanups.push(CleanupEntry {
            name: name.to_string(),
            func: Box::new(move || Box::pin(f())),
        });
    }

    pub async fn wait(&self) {
        let ctrl_c = tokio::signal::ctrl_c();
        
        #[cfg(unix)]
        let terminate = async {
            tokio::signal::unix::signal(tokio::signal::unix::SignalKind::terminate())
                .expect("failed to install signal handler")
                .recv()
                .await;
        };

        #[cfg(not(unix))]
        let terminate = std::future::pending::<()>();

        tokio::select! {
            _ = ctrl_c => {
                self.logger.info("Lifecycle: Received SIGINT. Initiating graceful shutdown...");
            }
            _ = terminate => {
                self.logger.info("Lifecycle: Received SIGTERM. Initiating graceful shutdown...");
            }
        }

        self.execute_cleanups().await;
    }

    pub async fn execute_cleanups(&self) {
        let mut cleanups = self.cleanups.lock().await;
        // Execute cleanups in reverse order (LIFO)
        while let Some(entry) = cleanups.pop() {
            self.logger.info(&format!("Lifecycle: Running cleanup '{}'...", entry.name));
            if let Err(e) = (entry.func)().await {
                self.logger.error(&format!("Lifecycle: Cleanup '{}' failed: {:?}", entry.name, e));
            }
        }
        self.logger.info("Lifecycle: Clean shutdown completed.");
    }
}

pub fn new_manager(logger: Option<Arc<dyn Logger>>) -> LifecycleManager {
    LifecycleManager::new(logger)
}
